src/common/Channel.cpp

```diff
@@ -231,6 +231,70 @@ void Channel::addMessagesAtStart(std::vector<MessagePtr> &_messages)
     }
 }
 
+void Channel::fillInMissingMessages(const std::vector<MessagePtr> &messages)
+{
+    auto snapshot = this->getMessageSnapshot();
+
+    std::unordered_set<QString> existingMessageIds;
+    existingMessageIds.reserve(snapshot.size());
+
+    // First, collect the ids of every message already present in the channel
+    for (auto &msg : snapshot)
+    {
+        if (msg->flags.has(MessageFlag::System) || msg->id.isEmpty())
+        {
+            continue;
+        }
+
+        existingMessageIds.insert(msg->id);
+    }
+
+    // Keep track of the last message in the channel. We need this value
+    // to allow concurrent appends to the end of the channel while still
+    // being able to insert just-loaded historical messages at the end
+    // in the correct place.
+    auto lastMsg = snapshot[snapshot.size() - 1];
+    for (auto &msg : messages)
+    {
+        // check if message already exists
+        if (existingMessageIds.count(msg->id) != 0)
+        {
+            continue;
+        }
+
+        bool insertedFlag = false;
+        for (auto &snapshotMsg : snapshot)
+        {
+            if (snapshotMsg->flags.has(MessageFlag::System))
+            {
+                continue;
+            }
+
+            if (msg->serverReceivedTime < snapshotMsg->serverReceivedTime)
+            {
+                // We found the first message that comes after the current message.
+                // Therefore, we can put the current message directly before.
+                this->messages_.insertBefore(snapshotMsg, msg);
+                insertedFlag = true;
+                break;
+            }
+        }
+
+        if (!insertedFlag)
+        {
+            // We never found a message already in the channel that came after
+            // the current message. Put it at the end and make sure to update
+            // which message is considered "the end".
+            this->messages_.insertAfter(lastMsg, msg);
+            lastMsg = msg;
+        }
+    }
+
+    // We only invoke a signal once at the end of filling all messages to
```

Comment:
If none of the messages received from the API were inserted, we still invoke the `filledInMessages`.
Would it make sense to use a variable to keep track of whether _any_ message has been inserted? And if not, skip invoking the signal?

Reply:
Good point 👍